[all] PSL branch prerequisites

bcarrez opened the pull request:
To review after #471 is merged (DONE) You can now review it :) 

This PR includes all changes and adds of the incoming [PSL] branch that can be merged directly. It includes some Python bindings, additional tests, and various cleanings.
Hopefully, it should not break anything.

 in BaseContext add variants of getObjects() which ease the syntax (eg: for(auto& i : getNodeObjects()){...} + corresponding tests. 

- in Node add variants of getNodeObjects() and getTreeObjects that does not force to ease the syntax+  corresponding tests. 

- in BaseLink add a setFlag method and a setPersistant...this makes things more consistent with BaseData.

- in Binding_Base add the methods: addNewData, addData, getData, getLink, getListOfDataFields, getListOfLinks. The function are documented with docstring and tested in test_BindingBase.py

- in Binding_Data add methods: isSet, isPersistant, setPersistant, hasParent, getParentPath, setValueString. The function are documented with docstring and tested in test_BindingData.py
 
- in Binding_Link add methods: setValueString, isPersistant, setPersistant. The function are documented with docstring and tested in test_BindingLink.py. 

- in Binding_SofaModule add methods: getAliasesFor. The function are documented with docstrings & tested in test_BindingSofa.py

- in SofaPython add functions: sendMessageFromException, getSofaFormattedStringFromException

- in SofaTest/_init_.py add ASSERT_* functions that generates real gtest failure. 

- in runSofa in QDataSimpleEdit & QLinkSimpleEdit correctly prevent writing when the data field is readOnly
______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

bcarrez committed:
[SofaPython] add some functions to bindings

bcarrez committed:
Merge commit '24f878820455d2c84ef10d44044792e0d9a342ca' into PSL_Preparation

bcarrez committed:
Merge remote-tracking branch 'origin/master' into PSL_Preparation

# Conflicts:
#	applications/plugins/SofaPython/Binding_SofaModule.cpp

bcarrez committed:
Backporting non-breaking modifications and tests from PSL branch

bcarrez committed:
fix Python_test build

damienmarchal commented:
OMG There is so much things to do to finalize this PR. 
I'm reviewing my own code but among things:
- add docstring,
- remove the todo before PR
- add tests to each new features.
- ...

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into HEAD

# Conflicts:
#	SofaKernel/framework/sofa/simulation/Node.h
#	applications/plugins/SofaPython/Binding_BaseContext.cpp
#	applications/plugins/SofaPython/python/SofaPython/__init__.py

damienmarchal committed:
[SofaKernel] Remove commented code.

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into PSL_Preparation

damienmarchal committed:
[SofaKernel] Add unittest for BaseContext

damienmarchal committed:
Merge remote-tracking branch 'origin/moveTestToSofaKernel2' into PSL_Preparation

damienmarchal committed:
[SofaKernel] Add/Fix few function "getObjects()" to BaseContext + test that validate them.

One of them allow the elegant (and RVO optimized) syntax of:
for(BaseObject* b : context.getObjects())
{
	...
}

damienmarchal committed:
[SofaKernel] Add few getNodeObjects & getTreeObjects in Node + corresponding tests.

damienmarchal committed:
[SofaKernel] restore the BaseTest.cpp

damienmarchal committed:
[SofaKernel/LocalStorage] Remove the include to BaseObject

Un-needed includes are bad. So we remove this one and fix the subsequents
files.

damienmarchal committed:
[SofaKernel/BaseSimulationTest] Add new helper functions to simulate a scene.

damienmarchal commented:
Ouch this PR is getting old. 
@bcarrez If you have time today it would be nice to finalize this PR so that we can move forward :)

damienmarchal commented:
Ouch I merged upstream/master & #471 so the diff is totally unreadable. Hope it will get better 
when 471 is merged.

damienmarchal committed:
[SofaKernel] Add tests for SingleLink

damienmarchal committed:
[SofaSimulationGraph] Improve BaseSimulationTest to use the right SceneLoader to load scene.

damienmarchal committed:
[SofaKernel] Remove the implicit loading of "SceneCreator" in BaseTest.

damienmarchal committed:
[SofaTest] Add new ASSERT_* api to the python module 'SofaTest'

The interest of these function is that they mimick the API of google
and correctly report the test failures in the tests results.

damienmarchal committed:
[SofaPython] Add new methods to Binding_Base.cpp

The methods are documented using docstring.

damienmarchal committed:
[SofaPython] Add 'empty' tests for BindingData, BindingBase, BindingLink, BindingSofa

The added tests files are working in a way close to the existing one in SofaTest
but:
- I wanted to implement tests for the recently added function needed by PSL without
refacting the whole python test framework right now. So it is currently not
factorized with the SofaTest approach

Unifying/cleaning everything will be a future work (once #471 is merged)

damienmarchal committed:
[SofaPython] Factorize the function pythonToSofaDataString in PythonMacro.cpp

damienmarchal committed:
[SofaPython] Add documentations & tests for Data, Link and SofaModule

damienmarchal committed:
[SofaTest] Adds additional ASSERT_* function into SofaTest/_init_.py

damienmarchal commented:
Hey @bcarrez I fixed all the things I wanted to do. We should move forward now.

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into PSL_Preparation

damienmarchal committed:
Merge remote-tracking branch 'upstream/master' into PSL_Preparation

damienmarchal committed:
[SofaKernel] Remove useless code.

damienmarchal commented:
Hi everyone, this this is an "all in one" branch that adds functions in Sofa 
before we can make a PR with the PSL plugin.

The PR is two month old...no one complains and now all the added function are tested & documented so it may be merged soon.

guparan commented:
Are you done with the rebase?

damienmarchal committed:
[SofaKernel/frameworkextra_test]Rename class in BaseContext_test that was wrong.

damienmarchal committed:
[SofaKernel/frameworkextra_test] Fix test to correctly load the required plugin

NB: There is a very interesting issue here about the required-plugin and
compilation links. It seems non trivial here to guarantee a full consistency
between the tests executed and the plugin states (i.e garanteeing that the
plugin is correctly reconstructed if the source code changed).

It is yet unclear to me how to handle that but this may be taken into account for the SofaNG.

damienmarchal committed:
[SofaPython] The pythonToSofaDataString was a wrong one not handling unicode. Now it does.

damienmarchal committed:
[SofaPython] Add *coding-utf8* at the beginning of python files.

damienmarchal commented:
I'm waiting the CI to finish but I think we are all done now.

damienmarchal commented:
@guparan, @fredroy, @bcarrez , @hugtalbot 
Hello, this branch is now synced with master, CI is ok, added functions have theirs tests...
So it is waiting.

damienmarchal commented:
No answer so "qui tacet consentire videtur" :)

jnbrunet commented on applications/plugins/SofaPython/Binding.cpp:
```diff
@@ -111,15 +110,16 @@ void bindSofaPythonModule()
 
-    /// Custom Exception to embed
+    /// Custom Exception
+    const char* name = "Sofa.SofaException";
```
Were those variables (name and doc) supposed to be use somewhere?

damienmarchal committed:
[SofaPython] Fix unused char*.

damienmarchal commented:
So I merge :)

guparan commented:
Houston we got a problem: https://ci.inria.fr/sofa-ci/job/ubuntu_clang-3.8_options-debug/545/testReport/
